### ISSUE TITLE
Do something about spamminess of log output

### ISSUE THREAD

**Issue by aboodman** [rocicorp/replicache-old#41]
This is a problem from even as far back as Turtle trying it out. Right now it's just super hard to follow what's going on in your app because Replicache is syncing so frequently and outputs so much.

But simply turning down the rate makes the experience unsatisfying of trying it out because you have to wait so long for changes to show up in the app.

In the long run, this obvious thing to do is push a notification to the device (or even a payload), but that's a much bigger discussion. Is there something we can do to make this less oft-putting?

It's challenging because everytime I turn this down I end up wanting the logs when debugging something.

I guess that probably the right thing to do is have a `logVerbosely` setting on `Replicache`. And when it's false, we only print out errors (and note that we shouldn't consider network problems to be errors).

See also #35

**Comment by phritz** [comment]
Part of the proposal in #35 is to A) define log levels and what is appropriate to log at each level and B) make a pass through the code ensuring that all calls to log are to a specific level and meet the policy. I think if we had a policy and if it were consistently applied then it would be easier to know the level at which to set logging in different environments.

If part of the problem is finding the right information in sea of log lines then a few things could help. First, a common pattern is to write log levels to different files (eg one has all the lines, one has only info and error, one has only error). This is really easy and enables you to notice the important signals with the option to zoom in on more detail if you want. Also, we could easily log specific things to specific log files, eg have a request/response log that just has that, have a noms log which just has its verbose output, etc.

The proposal in #35 is to have three log levels DEBUG, INFO, ERROR. We could have four (+VERBOSE) or two (-DEBUG) I don't care much, I just want to ensure we have a level that signals something is truly unexpected and a developer should look (ERROR in #35). 

I think #35 and this issue are basically the same thing. If you have specific examples of things that annoy you I bet we can find ways to make it better.

**Comment by phritz** [comment]
also note that to do most of the things suggested above we'd need to introduce something like the changes proposed in #35 . we can't just log directly to 'log'.

**Comment by aboodman** [comment]
Yeah i agree #35 is part of the solution to this. But even after that exists you still need a switch to turn it on and off via the bindings.

**Comment by aboodman** [comment]
Or rather to set the log level. That can be what this bug is about.

**Comment by phritz** [comment]
Is setting the log level from the app a specific instance of a more general issue, which is providing configuration from the app to the client? Eg customers are going to need to provide things like the batch endpoint url and maybe things like timeouts, retrys, etc.

**Comment by aboodman** [comment]
We already have a general solution to this problem, it's the `Replicache` constructor. We just need to add the field for this feature.

**Comment by phritz** [comment]
Constructor arguments as configuration are limited in the way you just pointed out: they can't be changed at runtime for example to adjust the log level.

**Comment by phritz** [comment]
what we'll do here: 
- have two levels
- have a field on replicache that controls which level we're at
- convert call sites in replicache but not noms

**Comment by aboodman** [comment]
At the risk of bringing up a contentious subject again, this is one of the first things people talk about when using Replicache (understandably) so I'm motivated thinking about it again.

Although categorizing levels will definitely help, if we see any output at all during sync, we're going to see it constant (until push is implemented).

I was thinking of trying something one of you mentioned (in addition to log levels):

* rate limit the same message getting repeated from sync to say ... ever 30s.
* unless the sync state changes in an interesting way (we send or receive new data, or connectivity changes).

**Comment by phritz** [comment]
If we have levels consistently used then we might not need the log every (N seconds or N occurrences) because of the second thing you say: we should only log when state changes in an interesting way. You have just described the INFO log level:

> ERROR: something truly unexpected has happened and a developer should go look. Example: https://github.com/rocicorp/diff-server/blob/a99baedd298425ab933e2c0cb62c0b393b371161/serve/service.go#L80
> INFO: an important change of state or information that is immediately useful to the developer. Example: https://github.com/rocicorp/diff-server/blob/a99baedd298425ab933e2c0cb62c0b393b371161/cmd/diffs/main.go#L127
> DEBUG: verbose information about what's happening that might be useful to a developer. Example: https://github.com/rocicorp/diff-server/blob/a99baedd298425ab933e2c0cb62c0b393b371161/serve/pull.go#L174

Take the log lines about sync for example. Sync starting or failing should not go into ERROR because we expect it to happen in the normal course of events. Nor should it go into INFO because it is not an important state change or immediately useful. It should go in DEBUG. The default setting for customers should be INFO (which shows INFO and ERROR) and so customers by default would not see sync starting or failing log lines. Sync completing successfully *with new state* should probably go in INFO bc it is an important change of state. Sync completing successfully with no new state is to me not interesting so I would put it at DEBUG.

We could provide the ability to override log level on a per feature or per package level. This could enable them to watch sync go for a while without being inundated from everywhere. Eg if INFO is the default level we could say something like this at startup: `log.Info("Syncing in the background every N seconds. Start with --debug-log=sync to see sync details logged.")`

**Comment by aboodman** [comment]
Yeah I decided I start with the log level thing and see how it looks after
that. Disagree that changing connectivity state isn't interesting though.
I'd put that in INFO.
